packages/api-db-mongodb/src/db/invoice.ts

```diff
@@ -0,0 +1,205 @@
+    const invoiceMap = Object.fromEntries(
+      invoices.map(({_id: id, ...invoice}) => [id, {id, ...invoice}])
+    )
+
+    return ids.map(id => invoiceMap[id] ?? null)
+  }
+
+  async getInvoices({
+    filter,
+    sort,
+    order,
+    cursor,
+    limit
+  }: GetInvoicesArgs): Promise<ConnectionResult<Invoice>> {
+    const limitCount = Math.min(limit.count, MaxResultsPerPage)
+    const sortDirection = limit.type === LimitType.First ? order : -order
+
+    const cursorData = cursor.type !== InputCursorType.None ? Cursor.from(cursor.data) : undefined
+
+    const expr =
+      order === SortOrder.Ascending
+        ? cursor.type === InputCursorType.After
+          ? '$gt'
+          : '$lt'
+        : cursor.type === InputCursorType.After
+        ? '$lt'
+        : '$gt'
+
+    const sortField = invoiceSortFieldForSort(sort)
+    const cursorFilter = cursorData
+      ? {
+          $or: [
+            {[sortField]: {[expr]: cursorData.date}},
+            {_id: {[expr]: cursorData.id}, [sortField]: cursorData.date}
+          ]
+        }
+      : {}
+
+    let textFilter: FilterQuery<any> = {}
+
+    // TODO: Rename to search
+    if (filter?.mail !== undefined) {
+      textFilter['$or'] = [{mail: {$regex: filter.mail, $options: 'i'}}]
+    }
+
+    const [totalCount, invoices] = await Promise.all([
+      this.invoices.countDocuments(textFilter, {
+        collation: {locale: this.locale, strength: 2}
+      } as MongoCountPreferences), // MongoCountPreferences doesn't include collation
+
+      this.invoices
+        .aggregate([], {collation: {locale: this.locale, strength: 2}})
+        .match(textFilter)
+        .match(cursorFilter)
+        .sort({[sortField]: sortDirection, _id: sortDirection})
+        .limit(limitCount + 1)
+        .toArray()
+    ])
+
+    const nodes = invoices.slice(0, limitCount)
+
+    if (limit.type === LimitType.Last) {
+      nodes.reverse()
+    }
+
+    const hasNextPage =
+      limit.type === LimitType.First
+        ? invoices.length > limitCount
+        : cursor.type === InputCursorType.Before
+        ? true
+        : false
+
+    const hasPreviousPage =
+      limit.type === LimitType.Last
+        ? invoices.length > limitCount
+        : cursor.type === InputCursorType.After
+        ? true
+        : false
+
+    const firstInvoice = nodes[0]
+    const lastInvoice = nodes[nodes.length - 1]
+
+    const startCursor = firstInvoice
+      ? new Cursor(firstInvoice._id, invoiceDateForSort(firstInvoice, sort)).toString()
+      : null
+
+    const endCursor = lastInvoice
+      ? new Cursor(lastInvoice._id, invoiceDateForSort(lastInvoice, sort)).toString()
+      : null
+
+    return {
+      nodes: nodes.map<Invoice>(({_id: id, ...invoice}) => ({id, ...invoice})),
+
+      pageInfo: {
+        startCursor,
+        endCursor,
+        hasNextPage,
+        hasPreviousPage
+      },
+
+      totalCount
+    }
+  }
+}
+
+function invoiceSortFieldForSort(sort: InvoiceSort) {
```

Comment:
Can we move it before it's being used?

Reply:
I recommend against it since all the files currently have the same structure.